drivers: serial: nrfx_uarte: Add support for device runtime PM and other minor improvements

nordic-krch opened the pull request:
PR adds:
- support for internal device runtime PM. Driver is internally calling `pm_device_runtime_get/put` when RX or TX activity is started. Additionally, initialization was reworked to call `pm_device_driver_init`.
- improvements in handling RX FIFO flushing. Additional workaround that attempts to detect if there are any bytes in FIFO by checking if RXDRDY event came after ENDRX. Also when PM is used, flushing happens only when UARTE is disabled and not on every RXTO.
- Minor: Fix compilation due to missing closing bracket, removal of unused element in the structure
- Making legacy shim the default for all targets.

nordic-krch committed:
drivers: serial: nrfx_uarte: Add runtime PM

Add runtime PM to the driver. When asynchronous or interrupt
driven API is used, there are 3 independent paths for enabling
the device: RX, TX and TX poll_out. TX poll_out requires special
handling because number of poll_out calls does not need to much
number of TXSTOPPED interrupts. To handle that special flag is
added. For standard RX and TX is simpler.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Use legacy shim by default

So far new platforms did not use legacy shim but it shall
be now a default uart shim for all platforms.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Remove unused element from a structure

Remove aborted flag from uarte_async_rx structure. Flag is not
used.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Add missing closing bracket

A code ifdefed by UARTE_BAUDRATE_RETENTION_WORKAROUND was missing a
closing bracket. It was missed because CI did not compile any target
that is applying this workaround and enables code compilation.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

bjarki-andreasen reviewed (dismissed):
looks good :)

nordic-krch committed:
drivers: serial: nrfx_uarte: Minor code size optimization

Optimize function which enables UARTE peripheral. It is called
only when interrupts are locked so data->flags does not require
atomic operation. Use standard logical operations so save few
bytes.

Simplify uarte_disable_locked.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

gmarull commented on drivers/serial/Kconfig.nrfx_uart_instance:
```diff
@@ -68,6 +68,7 @@ config UART_$(nrfx_uart_num)_NRF_ASYNC_LOW_POWER
 	depends on HAS_HW_NRF_UARTE$(nrfx_uart_num)
 	depends on UART_ASYNC_API
 	depends on UART_NRFX_UARTE_LEGACY_SHIM
+	depends on !PM_DEVICE
```
PM_DEVICE_RUNTIME implies this option, right?

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2185,9 +2185,9 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 		nrf_uarte_enable(uarte);
 #if UARTE_BAUDRATE_RETENTION_WORKAROUND
-		nrf_uarte_baudrate_set(get_uarte_instance(dev),
+	nrf_uarte_baudrate_set(get_uarte_instance(dev),
```
Seems that it was removed already.

anangl replied:
Well, it's still present in the first commit. But it's just a detail. We can live with this.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -328,34 +331,36 @@ static void uarte_nrfx_isr_int(const void *arg)
+	bool txstopped = nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_TXSTOPPED);
+
+	if (txstopped && (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME) ||
+	    (config->flags & UARTE_CFG_FLAG_LOW_POWER))) {
```
done

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -554,30 +559,29 @@ static int wait_tx_ready(const struct device *dev)
+#if UARTE_BAUDRATE_RETENTION_WORKAROUND
+	nrf_uarte_baudrate_set(uarte,
+			COND_CODE_1(CONFIG_UART_USE_RUNTIME_CONFIGURE,
+				(data->nrf_baudrate), (cfg->nrf_baudrate)));
```
done

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -827,6 +849,11 @@ static int uarte_nrfx_tx(const struct device *dev, const uint8_t *buf,
+		pm_device_runtime_get(dev);
+	}
+
+	nrf_uarte_int_enable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
```
I reworked it a bit. when `uart_tx()` is called we want to enable TXSTOPPED interrupt always (because it's async API). When poll_out is used then we want to enable TXSTOPPED interrupt only when PM or low power mode is used (to be able disable UARTE peripheral). I've moved enabling of TXSTOPPED to `poll_out` and removed one from `tx_start`.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1551,25 +1583,27 @@ static void txstopped_isr(const struct device *dev)
-	size_t amount = (data->async->tx.amount >= 0) ?
-			data->async->tx.amount : nrf_uarte_tx_amount_get(uarte);
-
 	irq_unlock(key);
```
done

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -1551,27 +1586,28 @@ static void txstopped_isr(const struct device *dev)
+	size_t amount = (data->async->tx.amount >= 0) ?
+			data->async->tx.amount : nrf_uarte_tx_amount_get(uarte);
+
+	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
+		nrf_uarte_int_disable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
+		if (data->flags & UARTE_FLAG_POLL_OUT) {
+			pm_device_runtime_put(dev);
+			data->flags &= ~UARTE_FLAG_POLL_OUT;
 		}
+	} else if (LOW_POWER_ENABLED(config)) {
+		nrf_uarte_int_disable(uarte, NRF_UARTE_INT_TXSTOPPED_MASK);
+		uarte_disable_locked(dev, UARTE_FLAG_LOW_POWER_TX, UARTE_FLAG_LOW_POWER_RX);
 	}
```
All of this is only for `uarte_nrfx_poll_out()`, right? Why not use the `UARTE_FLAG_POLL_OUT` flag to cover all this stuff here (and the related part of poll_out) and perform low-power disabling for a normal async transfer together with that `pm_device_runtime_put()` at the end of this function.
And similarly with `uarte_enable_locked()`, instead of calling it from `tx_start()`, why not call it together with `pm_device_runtime_get()` - separately in `start_tx_locked()` and `uarte_nrfx_poll_out()`? I think this would make things a bit clearer. And it will save some unnecessary low-power enabling and disabling when a transfer needs to be done in parts, through the TX cache buffer. There's one complication that I can see in such approach - if a TX transfer is pending because of an earlier poll_out, disabling would need to be skipped or re-enabling would need to be done before that pending transfer is started. This should not be that bad.

nordic-krch replied:
I did not apply it yet. Started to work on that but first draft shown slight code size increase. I'm also afraid about missing some corner cases (current implementation was passing all tests). Though, i agree that it would be slightly better to avoid disabling/re-enabling peripheral when cache is used. I will see tomorrow if i manage to add it in this PR.

anangl replied:
This can be improved in a follow-up PR.

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2156,93 +2172,70 @@ static void wait_for_tx_stopped(const struct device *dev)
-		if (ret < 0) {
-			return ret;
-		}
+	if (!LOW_POWER_ENABLED(cfg)) {
```
For `CONFIG_PM_DEVICE_RUNTIME`, low power flags should be either ignored here or not set at all in `dev->config`.

nordic-krch replied:
added check for runtime PM.

anangl commented on drivers/serial/Kconfig.nrfx_uart_instance:
```diff
@@ -68,7 +68,7 @@ config UART_$(nrfx_uart_num)_NRF_ASYNC_LOW_POWER
 	depends on UART_ASYNC_API
 	depends on UART_NRFX_UARTE_LEGACY_SHIM
-	default y
+	default y if !PM_DEVICE
```
This should be rather `PM_DEVICE_RUNTIME`. But maybe, instead of this, it would be better to mention in help that `PM_DEVICE_RUNTIME` overrides this setting, I mean that the instance will be appropriately disabled only if the runtime PM is enabled for it.

nordic-krch replied:
I've added note in help and do not set low power flag when runtime PM is used.

nordic-krch replied:
After more testing i came back to a conclusion that low power cannot be mixed with PM device. So I restored `if !PM_DEVICE` and note in  help. The reason is that when explicit device PM is used then we have two places where UART can be disabled (automatically in the interrupt and in the suspend action) and it is hard to manage. Given that we are going towards runtime device PM it does not make much sense to spend effort of feature that won't be used.

anangl replied:
Agreed. It's not worth spending more more time on this now. Excluding the old low-power mechanism entirely when PM is used is actually a "safe path" so let's take such approach.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2257,21 +2191,134 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 }
-#endif /* CONFIG_PM_DEVICE */
+
+static int uarte_config_init(const struct device *dev)
```
moved content of this function to the main init function.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2258,21 +2236,134 @@ static int uarte_nrfx_pm_action(const struct device *dev,
+	{
+		handle_endtx = true;
+	}
+#endif
```
done

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2258,21 +2236,134 @@ static int uarte_nrfx_pm_action(const struct device *dev,
+#endif
+	{
+		nrf_uarte_rx_buffer_set(uarte, cfg->poll_in_byte, 1);
+	}
```
done

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -2258,21 +2236,134 @@ static int uarte_nrfx_pm_action(const struct device *dev,
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (IS_ENABLED(CONFIG_PM_DEVICE)) {
+			uarte_pm_suspend(dev);
+		}
```
done

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -606,7 +626,7 @@ static void tx_start(const struct device *dev, const uint8_t *buf, size_t len)
 	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 
-#ifdef CONFIG_PM_DEVICE
+#ifdef	CONFIG_PM_DEVICE
```
removed.

nordic-krch commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -613,11 +614,13 @@ static void uarte_periph_enable(const struct device *dev)
 
-	if (atomic_or(&data->flags, act_mask) & sec_mask) {
+	data->flags |= act_mask;
+	if (already_active) {
```
removed `sec_mask`

anangl commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -252,6 +261,8 @@ struct uarte_nrfx_data {
 	(baudrate) == 1000000 ? NRF_UARTE_BAUDRATE_1000000 : 0)
 
+#define LOW_POWER_ENABLED(_config) \
+	(IS_ENABLED(UARTE_ANY_LOW_POWER) && (_config->flags & UARTE_CFG_FLAG_LOW_POWER))
```
Since the old low-power mechanism is not to be used with `PM_DEVICE`, maybe `!IS_ENABLED(CONFIG_PM_DEVICE)` should be added here, so that all related code could be optimized out?

nordic-krch replied:
added

nordic-krch committed:
drivers: serial: nrfx_uarte: Improve RX FIFO flushing workaround

Improved a workaround. When FIFO is not empty then FLUSHRX
task will generate RXSTARTED event to indicated that DMA transfer
was started. This property can be used to detect when FIFO was
not empty and workaroud a HW bug where RX.AMOUNT register is not
updated after flushing empty FIFO.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@anangl found much better (100% proof) workaround for RX FIFO flushing HW bug. I've added this workaround.

nordic-krch commented:
@dcpleung can you maybe take a look? It's vital to get that in before the freeze.
